operator/builtin/input/file/design.md

```diff
@@ -0,0 +1,146 @@
+
+1. Matching
+    1. Find files that match the `include` setting. The files are known only by their paths.
+    2. Discard any of these files that match the `exclude` setting.
+    3. As a special case, on the first poll cycle, a warning is printed if no files are matched. Execution continues regardless.
+2. Queueing
+    1. If the number of matched files is less than or equal to the maximum degree of concurrency, as defined by the `max_concurrent_files` setting, then no queueing occurs.
+    2. Else, queueing occurs, which means the following:
+        - Matched files are split into two sets, such that the first is small enough to respect `max_concurrent_files`, and the second contains the remaining files (called the queue).
+        - The current poll interval will begin processing the first set of files, just as if they were the only ones found during the matching phase.
+        - Subsequent poll cycles will pull matches off of the queue, until the queue is empty.
+        - The `max_concurrent_files` setting is respected at all times.
+        - TODO: Use a worker pool to process the queue as quickly as possible, rather than waiting for poll cycles to trigger batching.
+3. Opening
+    1. Each of the matched files is opened. Note:
+        - A small amount of time has passed since the file was matched.
+        - It is possible that is has been moved or deleted by this point.
+        - Only a minimum set of operations should occur between file matching and opening.
+        - If an error occurs while opening, it is logged.
+4. Fingerprinting
+    1. The first `N` bytes of each file are read. (See fingerprinting section above.)
+5. Exclusion
+    1. Empty files are closed immediately and discarded. (There is nothing to read.)
+    2. Fingerprints found in this batch are cross referenced against each other to detect duplicates. Duplicate files are closed immediately and discarded.
+        - In the vast majority of cases, this occurs during file rotation that uses the copy/truncate method. (See fingerprinting section above.)
+6. Reader Creation
+    1. Each file handle is wrapped into a `Reader` along with some metadata. (See Reader section above)
+        - During the creation of a `Reader`, the file's fingerprint is crossreferenced with previously known fingerprints.
+        - If a file's fingerprint matches one that has recently been seen, then metadata is copied over from the previous iteration of the Reader. Most importantly, the offset is accurately maintained in this way.
+        - If a file's fingerprint does not match any recently seen files, then its offset is initialized according to the `start_at` setting.
+7. Detection of "Lost" Files
+8. Consumption
+    1. Lost files are consumed. Since these files are thought to have been rotated to a location where we will not match them again, we will never see them again. 
+        - The best we can do is finish consuming their current contents.
```

Comment:
Unclear how we consume them if they are lost. Is it assumed that they are still open in a Reader but will be lost the moment we close them?

Reply:
Yes, that's the idea. 

By keeping the file open until the next poll cycle, we can compare two sets of "matched" files. Those that were found in the previous cycle but not in the current are considered lost. Several things may have happened, but in each case, the best we can do is try to read the file one last time.

If the file was deleted, then reading will be a nop. Typically, this would be due to rotation limits or ttl-based pruning, which means the older files have probably not been written to in a while, so we've fully consumed them.

If the file was moved, as is the case on k8s, then we can still read the remainder of it before closing. It's unlikely to be written to after being moved, so we'll have fully consumed it.

If the file was copied and truncated, we could potentially miss some lines. Rotation would have to follow essentially the same conditions as on k8s, but using copy/truncate. I'm not sure how likely this is to ever happen, but it does seem theoretically possible.